File: typescript/src/salesshift.ts
/**
 * SalesShift — the sales email service. Tracked sends through the org's
 * BYOK providers (tenant-node Go email worker preferred), suppression
 * gating, daily caps + warmup ramp, open/click tracking, and the
 * SendGrid-style Kafka event stream (salesshift.email.events).
 *
 * Endpoints (infinity control plane):
 *   POST /api/v1/salesshift/email/send
 *   GET  /api/v1/salesshift/emails
 *   GET  /api/v1/salesshift/stats
 * Endpoint (tenant node):
 *   GET  /api/v2/salesshift/email/health
 */

import type { Transport } from './transport.js';

export interface SendEmailInput {
  toEmail: string;
  subject: string;
  /** HTML body; merge tags like {{first_name}} resolve against the contact. */
  bodyHtml: string;
  firstName?: string;
  lastName?: string;
}

export interface SendEmailResult {
  success: boolean;
  status: string;      // sent | failed
  trackingId: string;
  provider: string;    // node-smtp | smtp | sendgrid | mailgun | platform | sink
  contactId: string;
  error?: string;
}

export interface TrackedEmail {
  id: string;
  toEmail: string;
  subject: string | null;
  status: string;
  provider: string | null;
  openCount: number;
  clickCount: number;
  sentAt: string | null;
}

export interface WorkerHealth {
  status: string;
  service: string;
  providers: string[];
  redisConnected: boolean;
  rateLimitDomain: number;
}

export class SalesShift {
  constructor(private t: Transport) {}

  /** Send one tracked email. Suppressed/unsubscribed recipients are rejected. */
  async sendEmail(input: SendEmailInput): Promise<SendEmailResult> {
    if (!input.toEmail || !input.subject || !input.bodyHtml) {
      throw new Error('salesshift.sendEmail: toEmail, subject and bodyHtml are required');
    }
    const res = await this.t.postJSON<Record<string, unknown>>('/api/v1/salesshift/email/send', {
      to_email: input.toEmail,
      subject: input.subject,
      body_html: input.bodyHtml,
      first_name: input.firstName,
      last_name: input.lastName,
    });
    const r = res.body ?? {};
    return {
      success: Boolean(r.success),
      status: String(r.status ?? ''),
      trackingId: String(r.tracking_id ?? ''),
      provider: String(r.provider ?? ''),
      contactId: String(r.contact_id ?? ''),
      error: r.error ? String(r.error) : undefined,
    };
  }

  /** List tracked emails, optionally filtered by status. */
  async listEmails(status?: string): Promise<TrackedEmail[]> {
    const path = `/api/v1/salesshift/emails${status ? `?status=${encodeURIComponent(status)}` : ''}`;
    const res = await this.t.get<{ data?: Record<string, unknown>[] }>(path);
    return (res.body?.data ?? []).map((m) => ({
      id: String(m.id ?? ''),
      toEmail: String(m.to_email ?? ''),
      subject: (m.subject as string) ?? null,
      status: String(m.status ?? ''),
      provider: (m.provider as string) ?? null,
      openCount: Number(m.open_count ?? 0),
      clickCount: Number(m.click_count ?? 0),
      sentAt: (m.sent_at as string) ?? null,
    }));
  }

  /** Live dashboard stats for the org. */
  async getStats(): Promise<Record<string, unknown>> {
    const res = await this.t.get<Record<string, unknown>>('/api/v1/salesshift/stats');
    return res.body ?? {};
  }

  /** Health of the tenant-node Go email worker. */
  async getWorkerHealth(): Promise<WorkerHealth> {
    const res = await this.t.get<Record<string, unknown>>('/api/v2/salesshift/email/health');
    const r = res.body ?? {};
    return {
      status: String(r.status ?? ''),
      service: String(r.service ?? ''),
      providers: (r.providers as string[]) ?? [],
      redisConnected: Boolean(r.redis_connected),
      rateLimitDomain: Number(r.rate_limit_domain ?? 0),
    };
  }
}

File: salesshift/salesshift.go
// Package salesshift wraps the SalesShift email service — tracked sends
// through the tenant's BYOK providers with suppression gating, daily caps,
// open/click tracking, and the SendGrid-style Kafka event stream.
//
// Endpoints (VxCloud control plane — FastAPI):
//
//	POST /api/v1/salesshift/email/send
//	GET  /api/v1/salesshift/emails
//	GET  /api/v1/salesshift/stats
//
// Endpoint (tenant node — Go email worker):
//
//	GET  /api/v2/salesshift/email/health
package salesshift

import (
	"context"
	"errors"
	"fmt"

	"github.com/prodxcloud/vxcloud/transport"
)

// Client — construct via c.SalesShift().
type Client struct {
	T           *transport.Transport
	VxCloudURL string
	NodeURL     string
}

// SendEmailInput is one tracked email. Subject/body accept merge tags
// ({{first_name}}, {{company_name}}) resolved against the contact record.
type SendEmailInput struct {
	ToEmail   string `json:"to_email"`
	Subject   string `json:"subject"`
	BodyHTML  string `json:"body_html"`
	FirstName string `json:"first_name,omitempty"`
	LastName  string `json:"last_name,omitempty"`
}

// SendEmailResult reports the delivery outcome and tracking handle.
type SendEmailResult struct {
	Success    bool   `json:"success"`
	Status     string `json:"status"` // sent | failed
	TrackingID string `json:"tracking_id"`
	Provider   string `json:"provider"` // node-smtp | smtp | sendgrid | mailgun | platform | sink
	ContactID  string `json:"contact_id"`
	Error      string `json:"error,omitempty"`
}

// SendEmail delivers one tracked email through the org's BYOK provider
// (tenant node worker preferred). Suppressed/unsubscribed recipients are
// rejected with an error — that gate is not optional.
func (c *Client) SendEmail(ctx context.Context, in SendEmailInput) (*SendEmailResult, error) {
	if in.ToEmail == "" || in.Subject == "" || in.BodyHTML == "" {
		return nil, errors.New("salesshift.SendEmail: ToEmail, Subject and BodyHTML are required")
	}
	url := transport.JoinURL(c.VxCloudURL, "/api/v1/salesshift/email/send")
	var out SendEmailResult
	if err := c.T.JSON(ctx, "salesshift.SendEmail", "POST", url, in, &out); err != nil {
		return nil, fmt.Errorf("salesshift.SendEmail: %w", err)
	}
	return &out, nil
}

// TrackedEmail is one row of the org's outbound email feed.
type TrackedEmail struct {
	ID         string `json:"id"`
	ToEmail    string `json:"to_email"`
	Subject    string `json:"subject"`
	Status     string `json:"status"`
	Provider   string `json:"provider"`
	OpenCount  int    `json:"open_count"`
	ClickCount int    `json:"click_count"`
	SentAt     string `json:"sent_at"`
}

// ListEmails returns tracked emails, optionally filtered by status
// (sent, opened, clicked, replied, bounced, unsubscribed, failed).
func (c *Client) ListEmails(ctx context.Context, status string) ([]TrackedEmail, error) {
	url := transport.JoinURL(c.VxCloudURL, "/api/v1/salesshift/emails")
	if status != "" {
		url += "?status=" + status
	}
	var raw struct {
		Success bool           `json:"success"`
		Data    []TrackedEmail `json:"data"`
	}
	if err := c.T.JSON(ctx, "salesshift.ListEmails", "GET", url, nil, &raw); err != nil {
		return nil, fmt.Errorf("salesshift.ListEmails: %w", err)
	}
	return raw.Data, nil
}

// Stats is the SalesShift dashboard aggregate.
type Stats struct {
	Contacts        int                    `json:"contacts"`
	Companies       int                    `json:"companies"`
	OpenDeals       int                    `json:"open_deals"`
	ActiveSequences int                    `json:"active_sequences"`
	EmailStats      map[string]int         `json:"email_stats"`
	Raw             map[string]interface{} `json:"-"`
}

// GetStats returns the org's live dashboard stats.
func (c *Client) GetStats(ctx context.Context) (*Stats, error) {
	url := transport.JoinURL(c.VxCloudURL, "/api/v1/salesshift/stats")
	var out Stats
	if err := c.T.JSON(ctx, "salesshift.GetStats", "GET", url, nil, &out); err != nil {
		return nil, fmt.Errorf("salesshift.GetStats: %w", err)
	}
	return &out, nil
}

// WorkerHealth reports the tenant node email worker's status.
type WorkerHealth struct {
	Status          string   `json:"status"`
	Service         string   `json:"service"`
	Providers       []string `json:"providers"`
	RedisConnected  bool     `json:"redis_connected"`
	RateLimitDomain int      `json:"rate_limit_domain"`
}

// GetWorkerHealth probes the Go email worker on the tenant node.
func (c *Client) GetWorkerHealth(ctx context.Context) (*WorkerHealth, error) {
	if c.NodeURL == "" {
		return nil, errors.New("salesshift.GetWorkerHealth: no tenant node resolved")
	}
	url := transport.JoinURL(c.NodeURL, "/api/v2/salesshift/email/health")
	var out WorkerHealth
	if err := c.T.JSON(ctx, "salesshift.GetWorkerHealth", "GET", url, nil, &out); err != nil {
		return nil, fmt.Errorf("salesshift.GetWorkerHealth: %w", err)
	}
	return &out, nil
}
